src/ngx_http_lua_tcpcong.c

```diff
@@ -0,0 +1,81 @@
+/*
+ * Copyright (C) Rain Li (blacktear23)
+ *
+ */
+
+#ifndef DDEBUG
+#define DDEBUG 0
+#endif
+#include "ddebug.h"
+
+
+#include "ngx_http_lua_tcpcong.h"
+#include "ngx_http_lua_util.h"
+#include "ngx_http_lua_api.h"
+
+
+int ngx_socket_set_tcp_congestion(ngx_socket_t s, const char* cong_name);
+static int ngx_http_lua_ngx_set_tcp_congestion(lua_State *L);
+
+
+#if (NGX_LINUX)
+
+int
+ngx_socket_set_tcp_congestion(ngx_socket_t s, const char* cong_name) {
+    u_char optval[16];
+    ngx_cpystrn(optval, (u_char *)cong_name, 16);
+    return setsockopt(s, IPPROTO_TCP, TCP_CONGESTION,
+                      (void *)optval, strlen((char *)optval));
+}
+
+#else
+
+int
+ngx_socket_set_tcp_congestion(ngx_socket_t s, const char* cong_name) {
+    return 0;
+}
+
+#endif
+
+/**
+ * Set TCP Congestion
+ * */
+static int
+ngx_http_lua_ngx_set_tcp_congestion(lua_State *L)
+{
+    ngx_http_request_t  *r;
+    int                 nargs;
+    int                 type;
+    int                 ret;
+    size_t              len;
+    const char          *p;
+
+    r = ngx_http_lua_get_request(L);
+    if (r == NULL) {
+        return luaL_error(L, "no request object found");
+    }
+
+    nargs = lua_gettop(L);
+    if (nargs != 1) {
+        return luaL_error(L, "attempt to pass %d arguments, bug accepted 1",
+                          nargs);
+    }
+    type = lua_type(L, 1);
+    if (type != LUA_TSTRING) {
+        return luaL_error(L, "require string parameter");
+    }
+    p = lua_tolstring(L, 1, &len);
+    ret = ngx_socket_set_tcp_congestion(r->connection->fd, p);
+    if (ret < 0) {
+        ngx_log_error(NGX_LOG_WARN, r->connection->log, 0,
+                      "set tcp congestion %s failed: %d", p, ret);
+    }
+    return 1;
+}
+
+void
+ngx_http_lua_inject_req_tcp_congestion_api(lua_State *L)
+{
+    lua_pushcfunction(L, ngx_http_lua_ngx_set_tcp_congestion);
+    lua_setfield(L, -2, "set_cong");
```

Comment:
I suggest use the current `setoption()` method instead of inventing a new method, as in

```lua
sock:setoption("congestion", "reno")
```

Reply:
I will try to implements this but will take more time.

Reply:
After I test my setoption implementation I found a problem: If use `ngx.req.socket(true)`, user should handle HTTP response by himself. If use `ngx.req.socket(false)` GET request will get error: `no body`. So if user just want to change request connection's TCP congestion `ngx.req.set_cong` will be the easy way to do it.

Reply:
@blacktear23 The "no body" error message is self-explanatory. The current request must have a request body to use this `ngx.req.socket(false)` API. Please read the official document more carefully.

Reply:
@blacktear23 No, `ngx.req.set_cong()` is a horrible API name. We should use the `setoption()` API instead. We definitely won't want to add a new Lua method for every new option we add, which will lead to a very bloated API and more memory and GC overhead.

Reply:
@agentzh what about add an API：ngx.req.setoption ？

Reply:
@blacktear23 Why? It should be an instance method instead of a static method. This is especially true for upstream cosockets created by `ngx.socket.tcp()`, for example.

Reply:
@agentzh One scenario is to use Lua change request connection's TCP congestion to improve TCP performance for file download, and then fall through Nginx basic logic. Such as the config below:

```
server {
    listen 80;
    location / {
        root /var/www/html;
        access_by_lua '
            ngx.req.setoption("congestion", "bbr")
        ';
    }
}
```

As you mentioned implement setoption for cosocket, after I get the socket by `ngx.req.socket` I cannot fall through Nginx basic logic. 

Is there I have some miss understanding on OpenResty?

Reply:
@blacktear23 That's a fair point. Then I suggest we add a static method to a new `ngx.req` Lua module in the lua-resty-core library:

https://github.com/openresty/lua-resty-core/

Just like its existing `ngx.ssl` and `ngx.re` Lua modules :)

We can just add a pure C API in this lua-nginx-module repo and invoke the C API through LuaJIT FFI from within the `ngx.req` module in lua-resty-core.